drive/src/drive/flags.rs

```diff
@@ -30,44 +30,620 @@
 use crate::error::Error;
 
+type EpochIndex = u16;
+
+type BaseEpoch = EpochIndex;
+
+type BytesAddedInEpoch = u32;
+
+type OwnerId = [u8; 32];
+
 /// Storage flags
-#[derive(Clone, Debug, Default, PartialEq)]
-pub struct StorageFlags {
-    /// Epoch
-    pub epoch: u16,
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum StorageFlags {
+    /// Single epoch
+    /// represented as byte 0
+    SingleEpoch(BaseEpoch),
+
+    /// Multi epoch
+    /// represented as byte 1
+    MultiEpoch(BaseEpoch, BTreeMap<EpochIndex, BytesAddedInEpoch>),
+
+    /// Single epoch owned
+    /// represented as byte 2
+    SingleEpochOwned(BaseEpoch, OwnerId),
+
+    /// Multi epoch owned
+    /// represented as byte 3
+    MultiEpochOwned(BaseEpoch, BTreeMap<EpochIndex, BytesAddedInEpoch>, OwnerId),
 }
 
 impl StorageFlags {
+    /// Create new single epoch storage flags
+    pub fn new_single_epoch(epoch: BaseEpoch, maybe_owner_id: Option<OwnerId>) -> Self {
+        match maybe_owner_id {
+            None => SingleEpoch(epoch),
+            Some(owner_id) => SingleEpochOwned(epoch, owner_id),
+        }
+    }
+
+    fn combine_owner_id<'a>(&'a self, rhs: &'a Self) -> Result<Option<&'a OwnerId>, Error> {
+        if let Some(our_owner_id) = self.owner_id() {
+            if let Some(other_owner_id) = rhs.owner_id() {
+                if our_owner_id != other_owner_id {
+                    return Err(Error::StorageFlags(
+                        StorageFlagsError::MergingStorageFlagsFromDifferentOwners(
+                            "can not merge from different owners",
+                        ),
+                    ));
+                }
+            }
+            Ok(Some(our_owner_id))
+        } else if let Some(other_owner_id) = rhs.owner_id() {
+            Ok(Some(other_owner_id))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn combine_non_base_epoch_bytes(
+        &self,
+        rhs: &Self,
+    ) -> Option<BTreeMap<EpochIndex, BytesAddedInEpoch>> {
+        if let Some(our_epoch_index_map) = self.epoch_index_map() {
+            if let Some(other_epoch_index_map) = rhs.epoch_index_map() {
+                let mut combined_index_map = our_epoch_index_map.clone();
+                other_epoch_index_map
+                    .iter()
+                    .for_each(|(epoch_index, bytes_added)| {
+                        let original_value = combined_index_map.remove(epoch_index);
+                        match original_value {
+                            None => combined_index_map.insert(*epoch_index, *bytes_added),
+                            Some(original_bytes) => combined_index_map
+                                .insert(*epoch_index, original_bytes + *bytes_added),
+                        };
+                    });
+                Some(combined_index_map)
+            } else {
+                Some(our_epoch_index_map.clone())
+            }
+        } else {
+            rhs.epoch_index_map().cloned()
+        }
```

Comment:
This is what I meant:
```suggestion
        // Combine index map if both index maps are present
        if let (Some(our_epoch_index_map), Some(other_epoch_index_map)) =
            (self.epoch_index_map(), rhs.epoch_index_map())
        {
            let mut combined_index_map = our_epoch_index_map.clone();
            other_epoch_index_map
                .iter()
                .for_each(|(epoch_index, bytes_added)| {
                    let original_value = combined_index_map.remove(epoch_index);
                    match original_value {
                        None => combined_index_map.insert(epoch_index.clone(), bytes_added.clone()),
                        Some(original_bytes) => combined_index_map.insert(
                            epoch_index.clone(),
                            original_bytes.clone() + bytes_added.clone(),
                        ),
                    };
                });

            return Some(combined_index_map);
        }

        self.epoch_index_map()
            .or_else(|| rhs.epoch_index_map())
            .cloned()
```

Reply:
I mean this is actually more checks and to me is less clear.

Reply:
I have opposite opinion. Since you have less conditions (nesting) it’s easier to read and understand. https://medium.com/swlh/return-early-pattern-3d18a41bba8

Reply:
But you the boss ))